---
title: "SimpLogo"
output: 
  github_document:
    dev: svg
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-"
)
```

<!-- badges: start -->
[![DOI](https://zenodo.org/badge/418649749.svg)](https://zenodo.org/badge/latestdoi/418649749)
<!-- badges: end -->

## Overview

The goal of SimpLogo is to develop a simplified, easy-to-read visualization for a large group of multiple amino acid sequence alignments that allows for convenient, simultaneous intergroup comparisons with minimized graphical clutter.

## Installation

You can install SimpLogo from Github using the devtools package:

``` r
library(devtools)
install_github("clayfos/SimpLogo",build_vignettes = TRUE)
```

## Example

SimpLogo utilizes the ggplot2 system to convert sequence alignments (in FASTA format) into pseudo-1D, color-coded rows. Because of this, all of the customization options offered by ggplot2 can be used to modify the resulting SimpLogo plot. Here is a basic example of analyzing a directory of FASTA alignments together.

### Input data

```{r example}
library(SimpLogo)

## use available files (.fa) in the sample_alignments directory
## we're only using 5 alignments for this example, but the 23 different CheW-containing architectures referenced in the publication can be found in the chew_alignments directory
## all alignments must be of equal length (for the example, all sequences were aligned simultaneously to the CheW-like Pfam HMM, then split into separate sequence groups)

## check filenames
seq.files <- list.files(path="sample_alignments/", 
                        all.files=TRUE, 
                        full.names=TRUE, 
                        include.dirs = FALSE,
                        pattern = "*fa")
seq.files

## you can manually specify alignment group names (for plotting)
## default takes them from the filenames
## make sure you order them correctly (check the filenames, like above)
groups <- c("CheA.I", "CheV.I", "CheW.IB", "CheW.IC", "CheW.II.1", "CheW.II.2")

## manually specify lineages/seq groups (for grouping during plotting)
## must have 1 assignment for every seq alignment/group being analyzed
lineage.assignments <- c("CheA","CheV","CheW","CheW","CheW","CheW")

results <- SimpLogo("sample_alignments/",
                    group.names = groups,
                    lineage.names = lineage.assignments)
```

The **results** object is a long-format dataframe containing the converted frequencies and blended colors in hexadecimal. This can be used as input for SimpLogoPlot().

```{r summary}
head(results)

## here, the group names defined above are stored in the "arch" column
```

### Plotting SimpLogo

SimpLogoPlot() uses ggplot2 to generate the plot objects (with geom_tiles). 

```{r plotting}
library(ggplot2)
library(Cairo)

plot <- SimpLogoPlot(results, plot.ic = TRUE)
```

It produces a list of 3 graphical results, the primary SimpLogo plot (the thing with the colored boxes) as a ggplot2 object, shown here: 


```{r fig2, fig.align = "center", dpi=300, echo = FALSE, out.width="100%"}
plot$primary.plot
```


A line graph of position-wise information content (in bits) as a ggplot2 object, shown here:
 
 
```{r fig3, fig.height = 2, fig.align = "center", dpi=300, echo = FALSE, out.width="100%"}
plot$ic.plot
```


And a final combined plot (for convenience) created using the plot_grid() function of cowplot, shown here:


```{r fig4, fig.align = "center", dpi=300, echo = FALSE, out.width="100%"}
plot$final.plot
```

### Manually editing individual graphs

You can edit the individual plots as ggplot2 objects by modifying the. Here we add a new overall title to our plot and change our axis text to green using theme [notice how we used the primary.plot object, the final.plot can't be edited due to the use of plot_grid()].

```{r fig5, fig.align = "center", dpi=300, out.width="100%"}
## now plot
plot$primary.plot + 
  labs(subtitle="this is an example title.") + 
  theme(axis.title.x.bottom =element_text(color="forestgreen"), 
        axis.text.x.bottom = element_text(color="forestgreen"), 
        axis.text.y.left = element_text(color="forestgreen"),
        plot.margin = margin(t=2)) ## to accomodate new title
```

### Recombining plots

If you decide to edit the ggplot2 objects, you can match the plot_grid() call in SimpLogoPlot() using the following:

```{r fig6, fig.align = "center", dpi=300, out.width="100%"}
final.plot <- cowplot::plot_grid(plot$ic.plot, plot$primary.plot,
                                 ncol = 1, align = 'v',
                                 axis = "lr", rel_heights = c(0.5, 1))
final.plot
```

### Renumbering alignment positions

If you'd like to easily renumber your alignment, starting at position 18 for example, you can use the following option when calling SimpLogoPlot() [or just modify it yourself by manipulating the ggplot2 object]

```{r fig7, fig.align = "center", dpi=300, out.width="100%"}
plot2 <- SimpLogoPlot(results, plot.ic = TRUE, position.start = 18)
plot2$final.plot
```

### Reordering sequence groups/architectures

You might want to reorder your architecture/sequence alignment groups. You can set the factor levels in the original results object to manually specify the order to ggplot2.

```{r fig9, fig.align = "center", dpi=300, out.width="100%"}
results2 <- results
results2$arch <- factor(results2$arch, levels = c("CheV.I","CheA.I","CheW.IB","CheW.IC","CheW.II.1","CheW.II.2"), ordered = TRUE) ## put CheV first, then CheA, then CheW
plot3 <- SimpLogoPlot(results2, plot.ic = TRUE, position.start = 18)
plot3$final.plot
```


## Output
The output of SimpLogo() is a dataframe with 10 fields:

* **color**: blended final color for main tile (in hexadecimal)
* **position**: position in alignment (starting from 1)
* **gap.freq**: frequency of gaps in alignment at position (1 = 100%)
* **info.content**: estimated information content (in bits) at position from alignment
* **top.type**: the most abundant residue type at the given position
* **secondary.type**: the second most abundant residue type at the given position
* **top.color**: idealized (100%) color for the most abundant residue type (in hexadecimal)
* **secondary.color**: idealized (100%) color for the second most abundant residue type (in hexadecimal)
* **arch**: sequence group assignment (taken from filenames if not s)
* **lineage**: higher level group assignment (useful for plotting)
* **residue**: residue number (used for plotting function)

The output of SimpLogoPlot() is a list containing 3 plot objects:

* **final.plot**: merged SimpLogo and IC plot (not a true ggplot2 object)
* **primary.plot**: main SimpLogo plot (editable ggplot2 object)
* **ic.plot**: line graph of estimated information content (editable ggplot2 object; NULL if plot.ic = FALSE)

## To do list
More customization options.
Possibly custom color palette options.
Add more complex instances in example section, including highlighting specific positions, etc.

## License information
This package is licensed under GPL3 - please refer to LICENSE.md for more details
